geoAssembler/geometry.py

```diff
@@ -70,17 +73,15 @@ def move_quad(self, quad, inc):
             quad (int): Quandrant number that is to be moved
             inc (collection): increment of the direction to be moved
         """
-        pos = Defaults.quad2index[self.detector_name][quad]
-        if len(inc) == 2:
-            inc = np.array(list(inc)+[0])
-        new_modules = [_move_mod(m, inc * self.pixel_size) if (pos <= i < pos + 4) else m
-                       for i, m in enumerate(self.modules)]
-        exgeom_cls = type(self.exgeom_obj)
-        self.exgeom_obj = exgeom_cls(new_modules)
-
-    @property
-    def _px_conv(self):
-        return self.pixel_size / self.unit
+        self.set_quad_offset(quad, self.quad_offsets[quad - 1] + inc)
+
+    def set_quad_offset(self, quad, offset):
+        self.quad_offsets[quad - 1] = offset
+        quad_offsets_m = self.quad_offsets * self.pixel_size
+        self.exgeom_obj = self.exgeom_obj_orig.offset(
+            # Repeat each quadrant offset 4 times to get offsets per module
+            np.repeat(quad_offsets_m, repeats=4)
+        )
```

Comment:
Looks like this is the bit causing the tests to fail. Haven't looked at extra-geom or geoAssembler internals in a while, but after skimming through them something like this seems to work:

```suggestion
        module_offsets_m = np.zeros((self.exgeom_obj.n_modules, 2))
        module_offsets_m[Defaults.quad2slice[self.detector_name][quad]] = quad_offsets_m
        self.exgeom_obj = self.exgeom_obj_orig.offset(module_offsets_m)
```

Far as I understand (this says little in a lot of words but I want to check I got it right in my mind), the geoAssembler `GeometryAssembler` stores the original geometry and an array of quadrant offsets. When the geometry gets shifted, the original geometry remains unmodified, the offsets are updated, and the shifted geometry is re-computed with the new offsets from the original geometry.

Newer versions of extra-geom have this offset method which can offset modules or tiles, so here I make an array of `module_offsets_m` from the `quad_offsets_m`, and pass that through to the offset method to shift things around.

Tests pass so either it works fine or the tests don't cover it :joy:

Reply:
Maybe you can take the opportunity and add codcov to the repository?

Reply:
Btw I just tried to understand whats going on and have absolutely no clue, mainly because I don't know the internals of extra-geom at all. So I can't really make any qualified comment.

Reply:
> Maybe you can take the opportunity and add codcov to the repository?

That is a good idea!

Reply:
> Btw I just tried to understand whats going on and have absolutely no clue, mainly because I don't know the internals of extra-geom at all. So I can't really make any qualified comment.

That's alright, thanks for having a look. Extra-geom internals changed so geoAssembler stopped working, Thomas started updating it to work again and then I stupidly upgraded the version of extra-geom in our environment and broke geoAssembler which is why I started working on it too.

Reply:
This was failing on CI because I made the PR before releasing EXtra-geom - I wanted to see how the new API worked downstream before committing to it. This was a useful exercise - I changed the API a bit in response to it. Now that it is released, I've rerun the tests, and... got an unrelated error from something in the GUI code. :disappointed: 

>  the geoAssembler GeometryAssembler stores the original geometry and an array of quadrant offsets. When the geometry gets shifted, the original geometry remains unmodified, the offsets are updated, and the shifted geometry is re-computed with the new offsets from the original geometry.

That's a pretty good summary, thanks!

My rationale for doing this is because EXtra-geom stores module positions as floating point numbers. If we repeatedly add/subtract floating point numbers, rounding errors can accumulate. If geoAssembler instead stores its offsets as integer numbers of pixels (the increments in which it works), and converts that into metres, there should only be a little bit of rounding error.

> Newer versions of extra-geom have this offset method which can offset modules or tiles, so here I make an array of module_offsets_m from the quad_offsets_m, and pass that through to the offset method to shift things around.

This is also what it was doing before - `np.repeat()` repeats each quadrant offset 4 times to get module offsets. But that does assume we have 4 modules per quadrant.

I think your suggested change would reset the other 3 quadrants to their original positions when you move any quadrant, though.

>  then I stupidly upgraded the version of extra-geom in our environment and broke geoAssembler

Upgrading EXtra-geom was a totally reasonable thing to do. I was aware of the incompatibility, but hadn't flagged it up - I guess I was hoping I could get geoAssembler fixed and upgrade them together.

> Maybe you can take the opportunity and add codcov to the repository?

:+1: But let's do it in a separate PR - this one is already big enough.

Reply:
>  Now that it is released, I've rerun the tests, and... got an unrelated error from something in the GUI code. disappointed

Ah, I just disabled the gui tests since they gave a segfault, assumed that was because of my.... non-standard configuration on my home computer(s)

> If we repeatedly add/subtract floating point numbers, rounding errors can accumulate.

Makes sense! Thanks for including comments in the code explaining that as well

> This is also what it was doing before - np.repeat() repeats each quadrant offset 4 times to get module offsets. But that does assume we have 4 modules per quadrant.

Aha, the `axis=0` bit is what was missing, that makes more sense than what I did :joy:

> I think your suggested change would reset the other 3 quadrants to their original positions when you move any quadrant, though.

Ooooh yes. @antarcticrainforest that's mainly what I meant by "or the tests don't cover it", even if you have coverage as in the lines get hit, if you do something creatively bad like I did in that case then the tests may pass even if the code doesn't work correctly.

@takluyver could you modify the tests to check move multiple quadrants around as well? I think currently it just moves one, and if somebody like me touches the code in the future then they may end up breaking things and not realising :sweat_smile: